ray-curator/ray_curator/stages/deduplication/fuzzy/connected_components.py

```diff
@@ -0,0 +1,209 @@
+
+    def weakly_connected_components(self, df: cudf.DataFrame, src_col: str, dst_col: str) -> None:
+        """Compute the weakly connected components of a graph.
+
+        This method loads a chunk of the graph, creates a cuGraph object, and
+        computes the weakly connected components using the MGGraph library.
+
+        Parameters
+        ----------
+        start: int
+            The start index of the chunk.
+        stop: int
+            The stop index of the chunk.
+        """
+
+        dg = cugraph.Graph(directed=False)
+
+        src_array = df[src_col]
+        dst_array = df[dst_col]
+
+        rhandle = ResourceHandle(self._raft_handle.getHandle())
+
+        graph_props = GraphProperties(
+            is_multigraph=False,  # dg.properties.multi_edge (what is multi_edge)
+            is_symmetric=not dg.graph_properties.directed,
+        )
+        logger.debug("Running graph creation")
+        plc_graph = MGGraph(
+            resource_handle=rhandle,
+            graph_properties=graph_props,
+            src_array=[src_array],
+            dst_array=[dst_array],
+            edge_id_array=None,
+            edge_type_array=None,
+            num_arrays=1,
+            store_transposed=False,
+            symmetrize=False,
+            do_expensive_check=False,
+            drop_multi_edges=True,
+        )
+        logger.debug("Running weakly connected components")
+        res = pylibcugraph_wcc(
+            resource_handle=rhandle,
+            graph=plc_graph,
+            offsets=None,
+            indices=None,
+            weights=None,
+            labels=None,
+            do_expensive_check=False,
+        )
+        logger.info("Computing weakly connected components completed successfully!")
+        return res
+
+    def process(self, task: FileGroupTask) -> FileGroupTask:
+        err_msg = "ConnectedComponentsStage only support process batch"
+        raise NotImplementedError(err_msg)
+
+    def process_batch(self, tasks: list[FileGroupTask]) -> list[FileGroupTask]:
+        """
+        Process an input file, compute minhashes, and write results to an output file.
+        Automatically adds a unique _curator_id field to each document if not present.
+
+        Parameters
+        ----------
+        infiles: str, list[str]
+            Path to input file (JSONL format) or list of paths
+        outfile: str
+            Path to output file (Parquet format)
+        columns: list, optional
+            Columns to read from input file
+        """
+        input_files = []
+        for task in tasks:
+            input_files.extend(task.data)
+        output_file = self.output_fs.sep.join([self.output_dir, f"{tasks[0].task_id}.parquet"])
+        edgelist_columns = [self.source_column, self.destination_column]
+        dfs = []
+        for input_file in input_files:
+            dfs.append(self.read_parquet(input_file, columns=edgelist_columns, **self.read_kwargs))
```

Comment:
Won't this be much slower than just doing `self.read_parquet(input_files)`? Why are we reading 1 by 1? Some memory thing I assume, if so how much is it, can you share? I might need to be mindful too.
FWIW I'm reading like ~300 files at a time using read_parquet and that's been fine, the biggest bottleneck has been the write_parquet step which seems to have way more of a memory overhead. see https://nvidia.slack.com/archives/CCWT6NJ0K/p1747335283494139

Reply:
It is much slower but comes with lower memory overhead. 
Here are results from reading a 8GB dataset split into 128 files.

```python
def read_one_at_a_time(files):
    dfs = []
    for file in files:
        dfs.append(cudf.read_parquet(file))
    df = cudf.concat(dfs, ignore_index=True)
    del(dfs)
    return df

def read_all(files):
    return cudf.read_parquet(files)
```

```
Read All: ~1s, 30GB peak memory usage
Read one at a time: ~9s, 16GB peak memory usage
```